feat(plugin-image-gallery): Open Image Plugin in Modal when user clicks Add "Images" in initial view (pe-52)

shn-srl opened the pull request:
**Linear Issue**: [Link](https://linear.app/serlo/issue/PE-52/popup-image-plugin)

**Demo Links**: 

[Editor Preview](https://frontend-git-pe-52-image-gallery-image-popup-serlo.vercel.app/___editor_preview?state=%7B%22plugin%22%3A%22rows%22%2C%22state%22%3A%5B%7B%22plugin%22%3A%22imageGallery%22%2C%22state%22%3A%7B%22images%22%3A%5B%5D%2C%22orderedIds%22%3A%22%22%7D%2C%22id%22%3A%22da744305-0622-45f7-ace6-d69a3c0f412d%22%7D%5D%7D)

[Create Article](https://frontend-git-pe-52-image-gallery-image-popup-serlo.vercel.app/entity/create/Article/1377)

shn-srl committed:
feat(plugin-image-gallery): Open popup with image when user clicks button in initial view

shn-srl committed:
wip: feat(plugin-image-gallery): Pass pluginId to ImageSelection screen

shn-srl commented on packages/editor/src/plugins/image/components/image-selection-screen.tsx:
```diff
@@ -19,6 +21,8 @@ export function ImageSelectionScreen(
   const { state, urlInputRef, setIsAButtonFocused } = props
   const { src, licence } = state
 
+  // const parentType = selectParentPluginType(store.getState(), pluginId)
```
Idea being: If the parentType is ImageGallery, then the Upload button can get passed a prop "multiple"

shn-srl replied:
(Perhaps not necessary, need to investigate the config and UploadHandler)

hejtful replied:
To save you some time on investigation: https://github.com/serlo/frontend/pull/4042

hejtful committed:
refactor(plugin-image): add multiple config prop

shn-srl committed:
Merge pull request #4042 from serlo/refactor/multiple-files-through-config

refactor(plugin-image): add multiple config prop

shn-srl committed:
feat(plugin-image-gallery): Add multipleUpload callback to imagePlugin config

shn-srl committed:
refactor(plugin-image-gallery): rename callback and make optional

hejtful commented on packages/editor/src/plugins/image/index.ts:
```diff
@@ -117,6 +117,8 @@ export type ImagePluginState = typeof imageState
 export type ImageProps = EditorPluginProps<ImagePluginState, ImageConfig>
 
 export interface ImagePluginConfig {
+  multiple?: boolean
+  onMultipleUploadCallback?: (files: File[]) => void
```
Not sure if you need two props. Instead of having `multiple` boolean, you can just check if the callback is defined.

shn-srl replied:
Done :heavy_check_mark:

hejtful commented on packages/editor/src/plugins/image-gallery/components/add-images.tsx:
```diff
@@ -1,11 +1,49 @@
+}
+
+export function AddImages(props: AddImagesProps) {
+  const { state, onAddImages } = props
+  const onClickAddImages = () => {
+    state.images.insert(0, createImagePlugin('', ''))
+    onAddImages()
+  }
```
Is there a need for splitting this logic across two components? Maybe just:
```suggestion
  const { onAddImages } = props
```
(continued in next comments)

Imo, splitting the logic across files makes it harder to understand, plus you need to pass more stuff around.

shn-srl replied:
Done :heavy_check_mark:

hejtful commented on packages/editor/src/plugins/image-gallery/components/add-images.tsx:
```diff
@@ -1,11 +1,49 @@
       <button
         className="mx-auto my-20 block rounded-md"
         data-qa="plugin-image-gallery-add-images-button"
+        onClick={() => onClickAddImages()}
```
```suggestion
        onClick={onAddImages}
```

hejtful commented on packages/editor/src/plugins/image-gallery/editor.tsx:
```diff
@@ -1,14 +1,60 @@
+      case ImageGalleryPluginViewType.INIT:
+        return (
+          <AddImages
+            onAddImages={() => {
+              setCurrentView(ImageGalleryPluginViewType.ADD_IMAGES)
+              setCurrentImageIndex(0)
+            }}
```
```suggestion
            onAddImages={() => {
              state.images.insert(0, createImagePlugin('', ''))
              setCurrentView(ImageGalleryPluginViewType.ADD_IMAGES)
              setCurrentImageIndex(0)
            }}
```

shn-srl replied:
Done :heavy_check_mark:

hejtful commented on packages/editor/src/plugins/image-gallery/components/add-images.tsx:
```diff
@@ -1,11 +1,49 @@
-export function AddImages() {
+import type { ImageGalleryProps } from '..'
+
+export function createImagePlugin(url: string, caption: string) {
+  return {
+    plugin: EditorPluginType.Image,
+    state: {
+      src: url,
+      caption: {
+        plugin: EditorPluginType.Text,
+        state: [
+          {
+            type: 'p',
+            children: [
+              {
+                text: caption,
+              },
+            ],
+          },
+        ],
+        id: uuidv4(),
+      },
+    },
+    id: uuidv4(),
+  }
+}
```
Move this into utils and name it `createEmptyImagePlugin`

shn-srl replied:
Done :heavy_check_mark:

hejtful commented on packages/editor/src/plugins/image-gallery/components/add-images.tsx:
```diff
@@ -1,11 +1,49 @@
+  }
+}
+
+interface AddImagesProps extends ImageGalleryProps {
+  onAddImages: () => void
+}
```
```suggestion
interface AddImagesProps {
  onAddImages: () => void
}
```

shn-srl replied:
Done :heavy_check_mark:

shn-srl committed:
refactor(plugin-image-gallery): Remove multiple prop from config, clean up code

shn-srl committed:
chore: remove unneeded change

shn-srl committed:
fix(plugin-image-gallery): Fix import

shn-srl commented on packages/editor/src/plugins/image/controls/upload-button.tsx:
```diff
@@ -55,11 +57,21 @@ export function UploadButton({
+              // If multiple files are allowed and more than one file is selected,
+              // call the onMultipleUploadCallback callback with the remaining files
+              if (config.onMultipleUploadCallback && filesArray.length > 1) {
+                config.onMultipleUploadCallback(filesArray.slice(1))
```
Todo: Find out how to insert new image plugins as children with these files assigned.

shn-srl committed:
feat(plugin-image-gallery): Enable multiple uploads

shn-srl committed:
refactor(plugin-image-gallery): remove pluginType

shn-srl committed:
feat(plugin-image-gallery): Show error if user tries to upload more than 8 images

elbotho commented on packages/editor/src/plugins/image-gallery/utils/create-empty-image-plugin.tsx:
```diff
@@ -0,0 +1,26 @@
+import { EditorPluginType } from '@editor/types/editor-plugin-type'
+import { v4 as uuidv4 } from 'uuid'
+
+export function createEmptyImagePlugin(url: string, caption: string) {
```
I'm confused about this one. Some questions:

- Why do you init the caption even though it's optional?
- Did you try not adding the uuids manually? I think they should be added if they are missing.

elbotho replied:
It's called `createEmptyImagePlugin`, why does it accept url and caption?

elbotho replied:
Did you try `state.images.insert(0, { plugin: EditorPluginType.Image })` above instead of this function?

shn-srl replied:
Removed.

elbotho commented on packages/editor/src/plugins/image/controls/upload-button.tsx:
Nice that you could reuse this file. 
We should probably have another string for multiple file uploads on the button:
<img width="295" alt="image" src="https://github.com/user-attachments/assets/bb30a3e3-b50f-459c-a7bf-df089904315c">

shn-srl replied:
Done :heavy_check_mark:

shn-srl committed:
refactor(plugin-image-gallery): remove superfluous helper file

shn-srl committed:
feat(plugin-image-gallery): Use different string for upload button when multiple allowed

shn-srl committed:
refactor(plugin-image-gallery): Use pluginStrings for error message

hejtful committed:
fix(store): ignore pending file upload action paths in serializableCheck

hejtful commented:
- When the modal is open, the toolbar says "Bild", not "Gallerie". It also disappears when the user clicks on it, which isn't a problem now, but might become a problem if that question mark icon from the design becomes clickable in the future.
- `DialogTitle` for a11y and avoiding the console error would be great.
- Unrelated to this PR perhaps, but still affecting this feature: the Image plugin in multimedia seems to have a higher z-index than both the modal dimmed background and the Image plugin within the modal.
<img width="878" alt="Screenshot 2024-08-22 at 15 14 09" src="https://github.com/user-attachments/assets/12959444-754d-41d3-9ecb-d0f8ccfe9476">

hejtful commented on packages/editor/src/plugins/image-gallery/components/fitting-modal.tsx:
```diff
@@ -0,0 +1,25 @@
+  onOpenChange: (open: boolean) => void
+}
+
+export function FittingModal({
```
Could you please explain why `ModalWithCloseButton` couldn't be used and a new modal component was needed?

hejtful commented on packages/editor/src/plugins/image-gallery/components/single-image-modal.tsx:
```diff
@@ -0,0 +1,36 @@
+  currentImageIndex: number
+  onAddImage: () => void
+  handleMultipleImageUpload: (files: File[]) => void
+}
```
```suggestion
interface SingleImageModalProps {
  currentImageState: ImageGalleryProps['state'][0]
  onAddImage: () => void
  handleMultipleImageUpload: (files: File[]) => void
}
```
This way, you have one prop less, and can skip line 15.

shn-srl replied:
Done

hejtful commented on packages/editor/src/plugins/image-gallery/editor.tsx:
```diff
@@ -1,14 +1,73 @@
+              setCurrentView(ImageGalleryPluginViewType.INIT)
+            }}
+            handleMultipleImageUpload={handleMultipleImageUpload}
+          />
```
```suggestion
          <SingleImageModal
            currentImageState={props.state.images[currentImageIndex]}
            onAddImage={() => {
              setCurrentView(ImageGalleryPluginViewType.INIT)
            }}
            handleMultipleImageUpload={handleMultipleImageUpload}
          />
```
Just to match the previous comment.

hejtful commented on packages/editor/src/plugins/image-gallery/editor.tsx:
```diff
@@ -1,14 +1,73 @@
+  INIT = 'INIT', // Initial state, no images
+  ADD_IMAGES = 'ADD_IMAGES', // Image selection screen
+  GALLERY = 'GALLERY', // Image grid
+}
```
Thanks for using an enum here! If comments are necessary to explain names, that usually tells me that names aren't good enough. Maybe consider:
```suggestion
enum ImageGalleryPluginView {
  EMPTY = 'EMPTY',
  SELECTING = 'SELECTING',
  GALLERY = 'GALLERY',
}
```
Or similar.

Also, couldn't find it being used anywhere else, so doesn't need to be exported.

shn-srl replied:
Done

hejtful commented on packages/editor/src/plugins/image-gallery/editor.tsx:
```diff
@@ -1,14 +1,73 @@
+    }
+  }
 
+  const renderContent = () => {
```
Having these `render*` function within a component is not something I prefer. If the JSX is too big, I prefer splitting the component into multiple smaller ones. If you do prefer using `render*` functions, however, please consider the [Order source files from abstract to concrete](https://github.com/serlo/documentation/wiki/Best-Practices#preference-order-source-files-from-abstract-to-concrete) best practice (also not something I personally prefer, but it's been a best practice in Serlo for a long time). So, using a function declaration instead of an arrow function and moving the declaration below the return statement would be preferred.

shn-srl replied:
Done

hejtful commented on packages/editor/src/plugins/image/controls/upload-button.tsx:
```diff
@@ -55,16 +58,33 @@ export function UploadButton({
         </span>
         <input
           type="file"
+          multiple={config.onMultipleUploadCallback ? true : false}
```
Super minor, just shorter:
```suggestion
          multiple={!!config.onMultipleUploadCallback}
```

hejtful reviewed (commented):
Mostly code style suggestions here:

shn-srl committed:
Update packages/editor/src/plugins/image/controls/upload-button.tsx

Co-authored-by: Vitomir Budimir <budimirvitomir@gmail.com>

shn-srl committed:
refactor(plugin-image-gallery): Dont use render function

shn-srl committed:
refactor(plugin-image-gallery): Better naming in View enum

shn-srl committed:
Merge branch 'pe-52-image-gallery-image-popup' of github.com:serlo/frontend into pe-52-image-gallery-image-popup

shn-srl committed:
refactor(plugin-image-gallery): Use currentImageState prop instead of index

shn-srl committed:
fix(plugin-image-gallery): Fix Dialog.title a11y issue

hejtful commented on packages/editor/src/plugins/image-gallery/components/fitting-modal.tsx:
```diff
@@ -0,0 +1,26 @@
+  return (
+    <Dialog.Root open={open} onOpenChange={onOpenChange}>
+      <Dialog.Portal>
+        <Dialog.Title className="sr-only" />
```
Same as here: https://github.com/serlo/frontend/pull/3986#discussion_r1705362937. It would be great to add an actual title for screen readers.

If you don't agree, I'd love to hear your arguments against it. If you do agree, it's weird that you added an empty title again.

shn-srl committed:
refactor(plugin-image-gallery): Use ModalWithCloseButton, remove FittingModal, improve styling

shn-srl committed:
chore(plugin-image-gallery): Add screen reader title to SingleImageModal

shn-srl commented:
> When the modal is open, the toolbar says "Bild", not "Gallerie". It also disappears when the user clicks on it, which isn't a problem now, but might become a problem if that question mark icon from the design becomes clickable in the future.

Fixed the modal layout so the toolbar is clickable.
"Bild" as title is the desired behaviour 
    

>     DialogTitle for a11y and avoiding the console error would be great.

Added

>    Unrelated to this PR perhaps, but still affecting this feature: the Image plugin in multimedia seems to have a higher z-index than both the modal dimmed background and the Image plugin within the modal.

Fixed by using ModalWithCloseButton


![image](https://github.com/user-attachments/assets/fce75cab-6a47-46af-9459-accb0421d4c6)

shn-srl committed:
feat(plugin-image-gallery): Use proper icon for Add Images button, add editor strings

shn-srl committed:
refactor(plugin-image-gallery): Fix colour of button text

elbotho commented:
> "Bild" as title is the desired behaviour

It says "Gallerie" in the [linked issue](https://linear.app/serlo/issue/PE-52/popup-image-plugin), did that change?

elbotho commented on packages/editor/src/editor-ui/assets/plugin-icons/image-gallery/icon-layered-images.svg:
Nice, thank you!

shn-srl commented:
> > "Bild" as title is the desired behaviour
> 
> It says "Gallerie" in the [linked issue](https://linear.app/serlo/issue/PE-52/popup-image-plugin), did that change?

You are correct, I was looking at the modal [here](https://linear.app/serlo/issue/PE-53/click-on-image-=-image-plugin-interface-popup)

The modal is functionally the same, but the title is different based on whether its showing the image selection screen or the rendered image

Fixed :heavy_check_mark:

shn-srl committed:
feat(plugin-image-gallery): Add title overwrite functionality to PluginToolbar, use dynamic title (Bild/Gallery) for Image plugin

hejtful reviewed (approved):
Looks good to me, thanks for all the fixes and adjustments
